app/services/manual_runner.rb

```diff
@@ -0,0 +1,46 @@
+class ManualRunner
+  def initialize(repo)
+    raise "Repo doesn't exist" unless Repo.exists?(repo.id)
+    @repo = repo
+    @octokit = Octokit::Client.new(access_token: Rails.application.secrets.github_api_token)
+  end
+
+  def run_last(commits_count)
+    run_paginated(commits_count)
+  end
+
+  private
+
+  def run_paginated(commits_count, page = 1)
+    unless (commits_count <= 0)
+      count_run = run_commits(page)
+      run_paginated(commits_count - count_run, page + 1)
+    end
+  end
+
+  def run_commits(page)
+    fetched_commits = @octokit.commits("#{@repo.organization.name}/#{@repo.name}", per_page: 100, page: page)
+    formatted_commits = format_commits(fetched_commits)
+    CommitsRunner.run(formatted_commits)
+
+    fetched_commits.count
+  end
+
+  def format_commits(commits)
+    commits.map do |commit|
+      {
+        sha: commit['sha'],
+        message: commit['commit']['message'],
+        repo: {
+          id: @repo.id,
+          name: @repo.name
+        },
+        url: commit['html_url'],
+        created_at: commit['commit']['author']['date'],
+        author: {
+          name: commit['commit']['author']['name']
+        }
+      }
```

Comment:
Given the amount of data that this method will have to handle, it may be worth trying to optimize it as much as you can. This is certainly going to be faster using something like:

~~~ruby
def format_commits(commits)
  commits.each do |commit|
    commit[:repo] = @repo
    commit.merge(commit['commit'])
    commit.symbolize_keys!
  end

  commits
end
~~~

You don't really care here if the objects are modified here. Also with the above example you don't have exactly the same API but not sure it's worth allocating a new hash to be able to do something like `commit[:url]` instead of `commit[:html_url]`.

I may be wrong though (both on the fact that you can modify objects without copying them and on the fact that may be faster) ; maybe you can write a tiny benchmark with `benchmark/ips` to see how fast it is and also check the number of allocated objects. What do you think ? 🙂